### DISCUSSION TITLE
HAL9000 on the IMSAI8080esp, with 80s speech synth

### DISCUSSION THREAD

**Discussion by donfroula** [thehighnibble/thehighnibble.github.io#166] [Show and tell]
I connected a serial port on my IMSAI8080esp replica and wrote a simple BASIC program in XYBASIC to read a text file from disc and output it a character at a time to the SIO Z80 port connected to the physical serial port. The speech synth then speaks the conversation HAL has with Dave Bowman while playing chess, just before he goes off the rails, from the film "2001: A Space Odyssey".

After each character, I check for an "XOFF" character, indicating the speech synth's buffer is full and to hold off sending more data. This accounts for the breaks in the text flow on the screen. The synth sends an "XON" character when it is ready to process more text. The program loops waiting for an XON after receiving the XOFF.

To add some visual bling, I also write each character to the Z80 port 0xFF, which is connected to the 8 "Programmed Output" LEDs on the front panel. The LEDs flash with the complement of the ASCII character code when data is sent to the synth.

https://youtu.be/yhAW5o629ks

**Comment by dmcnaugh** [comment]
Really cool. Nice work.
